sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java

```diff
@@ -94,4 +105,40 @@ public void processElement(ProcessContext c) {
             "RestoreOriginalTimestamps",
             ReifyTimestamps.<K, V>extractFromValues());
   }
+
+  /** Implementation of {@link #viaRandomKey()}. */
+  public static class ViaRandomKey<T> extends PTransform<PCollection<T>, PCollection<T>> {
+    private ViaRandomKey() {}
+
+    @Override
+    public PCollection<T> expand(PCollection<T> input) {
+      return input
+          .apply("Pair with random key", ParDo.of(new AssignShardFn<T>()))
+          .apply(Reshuffle.<Integer, T>of())
+          .apply(Values.<T>create());
+    }
+
+    private static class AssignShardFn<T> extends DoFn<T, KV<Integer, T>> {
+      private int shard;
+
+      @Setup
+      public void setup() {
+        shard = ThreadLocalRandom.current().nextInt();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        ++shard;
+        // Smear the shard into something more random-looking, to avoid issues
+        // with runners that don't properly hash the key being shuffled, but rely
+        // on it being random-looking. E.g. Spark takes the Java hashCode() of keys,
+        // which for Integer is a no-op and it is an issue:
+        // http://hydronitrogen.com/poor-hash-partitioning-of-timestamps-integers-and-longs-in-
+        // spark.html
+        // This hashing strategy is copied from com.google.common.collect.Hashing.smear().
+        int hashOfShard = 0x1b873593 * Integer.rotateLeft(shard * 0xcc9e2d51, 15);
```

Comment:
Any reason not to just call Hashing.smear?

Reply:
It's not public :(